### ISSUE TITLE
 Open issues regarding the Pixel local reconstruction on GPU

### ISSUE THREAD

**Issue by fwyzard** [cms-sw/cmssw#32483]
##  Open issues regarding the Pixel local reconstruction on GPU
  - [ ] make the CPU and GPU buffers used for the pixel local reconstruction configurable at runtime

**Comment by fwyzard** [comment]
## make the CPU and GPU buffers used for the pixel local reconstruction configurable at runtime

See https://github.com/cms-sw/cmssw/pull/31721/files#r521009826 for an initial discussion.

The dimensions defined in [CUDADataFormats/SiPixelCluster/interface/gpuClusteringConstants.h](https://github.com/cms-patatrack/cmssw/tree/master/CUDADataFormats/SiPixelCluster/interface/gpuClusteringConstants.h) should eventually be changes to accommodate the Phase 2 and Heavy Ions conditions.

This is possible in different ways:
  - making 2-3 variants of the classes (*e.g.* templating them) for few well-defined use cases, and pick a different implementation at runtime based on the configuration
  - making the buffer sizes dependent on the job configuration, and allocate them at module construction time
  - making the buffers fully resizable at runtime (or at least event by event) (unlikely to be useful)
  - *etc.*

Some preliminary work was done in https://github.com/cms-patatrack/cmssw/pull/545, https://github.com/cms-patatrack/cmssw/pull/583, https://github.com/cms-patatrack/cmssw/pull/590 .

A likely requirement is a more user-friendly SoA interface (see https://github.com/cms-patatrack/cmssw/issues/272).

**Comment by fwyzard** [comment]
assign reconstruction

**Comment by fwyzard** [comment]
assign heterogeneous

**Comment by fwyzard** [comment]
@tsusa @mmusich @VinInn FYI

**Comment by fwyzard** [comment]
## Reduce the number of dedicated namespacess for Pixel reconstruction on GPU
  - `namespace gpuClustering` is used in
      - [CUDADataFormats/SiPixelCluster/interface/gpuClusteringConstants.h](https://github.com/cms-patatrack/cmssw/tree/master/CUDADataFormats/SiPixelCluster/interface/gpuClusteringConstants.h)
      - [RecoLocalTracker/SiPixelClusterizer/plugins/gpuClusterChargeCut.h](https://github.com/cms-patatrack/cmssw/tree/master/RecoLocalTracker/SiPixelClusterizer/plugins/gpuClusterChargeCut.h)
      - [RecoLocalTracker/SiPixelClusterizer/plugins/gpuClustering.h](https://github.com/cms-patatrack/cmssw/tree/master/RecoLocalTracker/SiPixelClusterizer/plugins/gpuClustering.h)
  - `namespace pixelGPUConstants` is used in
      - [CUDADataFormats/SiPixelCluster/interface/gpuClusteringConstants.h](https://github.com/cms-patatrack/cmssw/tree/master/CUDADataFormats/SiPixelCluster/interface/gpuClusteringConstants.h)

**Comment by fwyzard** [comment]
## Add more comments to the code

  - in [TrackingRecHit2DHeterogeneous.h](https://github.com/cms-patatrack/cmssw/blob/6a192beda960630de730923b10a44a5a1da5e77d/CUDADataFormats/TrackingRecHit/interface/TrackingRecHit2DHeterogeneous.h#L45-L46):
    ```c++
      static constexpr uint32_t n16 = 4;
      static constexpr uint32_t n32 = 9;
    ```

**Comment by fwyzard** [comment]
## Use `gpuClustering::maxNumModules` where appropriate

Various pieces of code use `2000` or `2001` by value; figure out which uses can and should be replace by `gpuClustering::maxNumModules` or `gpuClustering::maxNumModules + 1`:

  - CUDADataFormats/SiPixelCluster/interface/gpuClusteringConstants.h:
    ```c++
    constexpr uint16_t maxNumModules = 2000;
    ```

  - CUDADataFormats/TrackingRecHit/src/TrackingRecHit2DHeterogeneous.cc:
    ```c++
    auto ret = cms::cuda::make_host_unique<uint32_t[]>(2001, stream);
    cudaCheck(cudaMemcpyAsync(ret.get(), m_hitsModuleStart, 4 * 2001, cudaMemcpyDefault, stream));
    ```

  - CondFormats/SiPixelObjects/interface/SiPixelGainForHLTonGPU.h:
    ```c++
    std::pair<Range, int> rangeAndCols[2000];
    ```

  - RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelDigisClustersFromSoA.cc:
    ```c++
    outputClusters->reserve(2000, nDigis / 4);
        auto clusterThreshold = (layer == 1) ? 2000 : 4000;
    ```

  - RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelRawToClusterGPUKernel.h:
    ```c++
    const uint32_t MAX_WORD = 2000;
    ```

  - RecoLocalTracker/SiPixelClusterizer/test/gpuClustering_t.h:
    ```c++
    int numElements = 256 * 2000;
    ```

  - RecoLocalTracker/SiPixelRecHits/plugins/SiPixelRecHitSoAFromLegacy.cc:
    ```c++
    assert(2001 == clusInModule_.size());
    assert(0 == clusInModule_[2000]);
        assert(gind < 2000);
    assert(numberOfClusters == int(hitsModuleStart[2000]));
        legacyOutput->reserve(2000, numberOfClusters);
        assert(gind < 2000);
    ```

  - RecoPixelVertexing/PixelTriplets/plugins/gpuPixelDoubletsAlgos.h:
    ```c++
        if (mi > 2000)
            if (mo > 2000)
    ```

  - SimTracker/TrackerHitAssociation/plugins/ClusterSLOnGPU.cu
    ```c++
    assert(id < 2000);
    assert(cl < 2000 * blockDim.x);
    ```

**Comment by fwyzard** [comment]
## Use `phase1PixelTopology::numberOfLayers` where appropriate

In
  - Geometry/TrackerGeometryBuilder/test/phase1PixelTopology_t.cpp
  - RecoLocalTracker/SiPixelRecHits/plugins/SiPixelRecHitSoAFromLegacy.cc

use `phase1PixelTopology::numberOfLayers` (`+ 1`) in place of `10` (`11`).

**Comment by VinInn** [comment]
not these one
```
RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelDigisClustersFromSoA.cc:

   /// this is an algorithm constant defined by Danek
    auto clusterThreshold = (layer == 1) ? 2000 : 4000;


RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelRawToClusterGPUKernel.h:

// this is a property of FED/ROC and already proposed to be increased for HI
const uint32_t MAX_WORD = 2000;
```

**Comment by fwyzard** [comment]
## Make the pixel cluster thresholds configurable

`RecoLocalTracker/SiPixelClusterizer/plugins/gpuClusterChargeCut.h` and `RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelDigisClustersFromSoA.cc` use hard-coded cuts of `2000` and `4000`, while the CPU code at `RecoLocalTracker/SiPixelClusterizer/plugins/PixelThresholdClusterizer.cc` takes the same values from the runtime configuration.

The GPU code could be updated to do the same.

**Comment by czangela** [comment]
I would like to reiterate and start a conversation on the matter of namespaces, listed in [this previous comment](https://github.com/cms-sw/cmssw/issues/32483#issuecomment-744671102).

Some previous conversations about this went on at #31721 and can be found at [comment 1](https://github.com/cms-sw/cmssw/pull/31721#discussion_r521009826), [comment 2](https://github.com/cms-sw/cmssw/pull/31721#discussion_r527377013). Please add more if you know of any. (threads, comments, hypernews, anything)

As far as I understand, (correct me @mmusich if I'm wrong) one namespace should do.

Ideas on what this should be are welcome.

**Comment by VinInn** [comment]
namespaces are used to group "items" no more no less than structs or classes. The same OOD criteria apply.
One structural difference is that declarations in namespaces may span multiple files.
```phase1PixelTopology``` is the kind of entity that in the old days would have been modeled using a struct.
I find a namespace more appropriate (and in CMSSW we had campaigns to move "struct" to "namespace").
Still, as we deplorate schizophrenic classes, we shall not, in my opinion, clusterize everything is a single namespace just for the sake of "economy".
In summary: just use your best OOD judgment.

**Comment by jpata** [comment]
- [ ] std::clamp in gpuClusterTracksIterative.h, gpuClusterTracksDBSCAN.h https://github.com/cms-sw/cmssw/pull/31723#discussion_r605817407
- [ ] repeated constants in gpuVertexFinderImpl.h, https://github.com/cms-sw/cmssw/pull/31723#discussion_r605820701

**Comment by mmusich** [comment]
@jpata, can you please move the open issues connected to pixel tracking and vertexing to a separate issue?
As far as I can tell (based on the title) this issue is reserved to pixel local reconstruction.
Thanks.
